quickjs/src/main/jni/JsMethodProxy.cpp

```diff
@@ -18,53 +18,119 @@
 #include <algorithm>
 #include "Context.h"
 
-JsMethodProxy::JsMethodProxy(JNIEnv *env, const char *name, jobject method)
-    : name(name), methodId(env->FromReflectedMethod(method)) {
+namespace {
+
+std::string getName(JNIEnv *env, jclass javaClass) {
+  const jmethodID method =
+      env->GetMethodID(env->GetObjectClass(javaClass), "getName", "()Ljava/lang/String;");
+  auto javaString = static_cast<jstring>(env->CallObjectMethod(javaClass, method));
+  const auto s = env->GetStringUTFChars(javaString, nullptr);
+
+  std::string str(s);
+  env->ReleaseStringUTFChars(javaString, s);
+  env->DeleteLocalRef(javaString);
+  return str;
+}
+
+JsMethodProxy::ArgumentLoader
+getJavaToJsConverter(const Context* context, jclass type, bool boxed) {
+  const auto typeName = getName(context->env, type);
+  if (typeName == "java.lang.String") {
+    return [](const Context* c, jvalue v) {
+      if (!v.l) return JS_NULL;
+      const auto s = c->env->GetStringUTFChars(static_cast<jstring>(v.l), nullptr);
+      auto jsString = JS_NewString(c->jsContext, s);
+      c->env->ReleaseStringUTFChars(static_cast<jstring>(v.l), s);
+      return jsString;
+    };
+  } else if (typeName == "java.lang.Double" || (typeName == "double" && boxed)) {
+    return [](const Context *c, jvalue v) {
+      return v.l != nullptr
+             ? JS_NewFloat64(c->jsContext, c->env->CallDoubleMethod(v.l, c->doubleGetValue))
+             : JS_NULL;
+    };
+  } else if (typeName == "java.lang.Integer" || (typeName == "int" && boxed)) {
+    return [](const Context* c, jvalue v) {
+      return v.l != nullptr
+             ? JS_NewInt32(c->jsContext, c->env->CallIntMethod(v.l, c->integerGetValue))
+             : JS_NULL;
+    };
+  } else if (typeName == "java.lang.Boolean" || (typeName == "boolean" && boxed)) {
+    return [](const Context* c, jvalue v) {
+      return v.l != nullptr
+             ? JS_NewBool(c->jsContext, c->env->CallBooleanMethod(v.l, c->booleanGetValue))
+             : JS_NULL;
+    };
+  } else if (typeName == "double") {
+    return [](const Context* c, jvalue v) {
+      return JS_NewFloat64(c->jsContext, v.d);
+    };
+  } else if (typeName == "int") {
+    return [](const Context* c, jvalue v) {
+      return JS_NewInt32(c->jsContext, v.i);
+    };
+  } else if (typeName == "boolean") {
+    return [](const Context* c, jvalue v) {
+      return JS_NewBool(c->jsContext, v.z);
+    };
+  } else {
+    return [](const Context*, jvalue) { return JS_NULL; };
```

Comment:
Is it going to be handled via an exception in the future?

Reply:
yeah, I'll add a TODO